nasa/fprime
# FPP v2.0 integration

## Change Description

This PR integrates FPP v2.0, which adds support for C++ generation of unit test harness classes, with F Prime.

Unit tests for the FPP unit test autocoder have been added to the existing tests for FPP components.

## Changes to Generated C++ Code for Unit Test Harnesses

### General
- Prepend the component name to the generated file names, e.g. `[ComponentName]TesterBase.cpp` instead of `TesterBase.cpp`. This allows the generation of test harness classes for multiple components within the same directory.

### TesterBase classes
- Reorganize class members (follows the same pattern as the reorganization of class members for component base classes, see #2103)
- Change `typedef struct {...} [name];` to `struct [name] {...};`
- Move initialization of `m_param_[paramName]_valid` variable members from the `init` function to initializer lists in the constructor
- Fix issue where `[portName]_static` and dispatch function implementations for special ports are not generated when the corresponding special port constructs do not appear (e.g. a component contains command ports, but no commands)
- Change `sizeof([serializableType])` to `[serializableType]::SERIALIZED_SIZE` where appropriate in the `dispatchEvents` function

## To do before merging

- [x] Merge unit test code generation into FPP
- [ ] Integrate unit test code generation with fprime tools
- [ ] Create FPP pre-release to test PR
- [ ] Create FPP release v2.0
